src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1049,113 +1045,156 @@ public bool TakeStartLine(ReadableBuffer buffer, out ReadCursor consumed, out Re
-                    {
-                        RejectRequestLine(start, end);
-                    }
-                }
-            }
-            else
-            {
-                methodEnd += method.Length;
-            }
-
             var needDecode = false;
-            var pathBegin = methodEnd + 1;
-            var pathToEndSpan = span.Slice(pathBegin, span.Length - pathBegin);
-            pathBegin = 0;
-
-            // TODO: IndexOfAny
-            var spaceIndex = pathToEndSpan.IndexOf(ByteSpace);
-            var questionMarkIndex = pathToEndSpan.IndexOf(ByteQuestionMark);
-            var percentageIndex = pathToEndSpan.IndexOf(BytePercentage);
-
-            var pathEnd = MinNonZero(spaceIndex, questionMarkIndex, percentageIndex);
-
-            if (spaceIndex == -1 && questionMarkIndex == -1 && percentageIndex == -1)
-            {
-                RejectRequestLine(start, end);
-            }
-            else if (percentageIndex != -1)
-            {
-                needDecode = true;
-
-                pathEnd = MinNonZero(spaceIndex, questionMarkIndex);
-                if (questionMarkIndex == -1 && spaceIndex == -1)
-                {
-                    RejectRequestLine(start, end);
-                }
-            }
-
+            var pathStart = -1;
+            var queryStart = -1;
+            var queryEnd = -1;
+            var pathEnd = -1;
+            var versionStart = -1;
             var queryString = "";
-            var queryEnd = pathEnd;
-            if (questionMarkIndex != -1)
-            {
-                queryEnd = spaceIndex;
-                if (spaceIndex == -1)
-                {
-                    RejectRequestLine(start, end);
-                }
+            var httpVersion = "";
+            var method = "";
+            var state = StartLineState.ExpectKnownMethod;
 
-                queryString = pathToEndSpan.Slice(pathEnd, queryEnd - pathEnd).GetAsciiString();
-            }
-
-            if (pathBegin == pathEnd)
-            {
-                RejectRequestLine(start, end);
-            }
-
-            var versionBegin = queryEnd + 1;
-            var versionToEndSpan = pathToEndSpan.Slice(versionBegin, pathToEndSpan.Length - versionBegin);
-            versionBegin = 0;
-            var versionEnd = versionToEndSpan.IndexOf(ByteCR);
-
-            if (versionEnd == -1)
+            fixed (byte* data = &span.DangerousGetPinnableReference())
             {
-                RejectRequestLine(start, end);
-            }
-
-            if (!versionToEndSpan.GetKnownVersion(out string httpVersion))
-            {
-                httpVersion = versionToEndSpan.Slice(0, versionEnd).GetAsciiStringEscaped();
-
-                if (httpVersion == string.Empty)
-                {
-                    RejectRequestLine(start, end);
-                }
-                else
+                var length = span.Length;
+                for (var i = 0; i < length; i++)
```

Comment:
yes, it's actually parsing byte by byte right now, there just happens to be cases where it jumps forward (fast method and version).